feat: check node engines version in package.json

t-mangoe opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!--- To help with semantic versioning the PR title should start with one of the conventional commit types. -->
<!--- The conventional commit types for Semantic PR are: feat, fix, docs, style, refactor, perf, test, build, ci, chore, revert -->

#### Description
<!--- Describe your changes in detail -->
I added codes which check node.js engines version written in package.json

#### Motivation and Context
<!--- Why is this change required? hat problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
Closes #1605

#### How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, tests ran to see how -->
<!--- your change affects other areas of the code, etc. -->
- [ ] I have tested using **MacOS**
- [x] I have tested using **Linux**
- [ ] I have tested using **Windows**

#### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] I have updated the documentation accordingly.
- [x] I have updated the tests accordingly.
 -> I added two tests.

andytom commented on Cargo.toml:
```diff
@@ -30,16 +30,16 @@ tls-vendored = ["native-tls/vendored"]
 [dependencies]
 clap = "2.33.3"
 ansi_term = "0.12.1"
-dirs-next = "2.0.0"
```
Can you revert the downgrades to these libraries?

andytom commented on src/modules/nodejs.rs:
```diff
@@ -59,12 +73,45 @@ pub fn module<'a>(context: &'a Context) -> Option<Module<'a>> {
 }
 
+fn get_engines_version(base_dir: &PathBuf) -> Option<String> {
+    if let Ok(json_str) = utils::read_file(base_dir.join("package.json")) {
+        let package_json: json::Value = json::from_str(&json_str).ok()?;
+        let raw_version = package_json.get("engines")?.get("node")?.as_str()?;
+        Some(raw_version.to_string())
+    } else {
+        None
+    }
```
Not sure if it matters but you could simplify this a little bit to:
```suggestion
    let json_str = utils::read_file(base_dir.join("package.json")).ok()?;
    let package_json: json::Value = json::from_str(&json_str).ok()?;
    let raw_version = package_json.get("engines")?.get("node")?.as_str()?;
    Some(raw_version.to_string())
```

andytom commented on src/modules/nodejs.rs:
```diff
@@ -59,12 +73,45 @@ pub fn module<'a>(context: &'a Context) -> Option<Module<'a>> {
+    }
+}
+
+fn check_engines_version(_nodejs_version: &str, _engines_version: Option<String>) -> bool {
```
It's generally better practice not to prefix variables with a `_` as it will stop the compiler from warning if the variable is not used.

```suggestion
fn check_engines_version(nodejs_version: &str, engines_version: Option<String>) -> bool {
```

andytom commented on src/modules/nodejs.rs:
```diff
@@ -31,14 +37,22 @@ pub fn module<'a>(context: &'a Context) -> Option<Module<'a>> {
     let mut module = context.new_module("nodejs");
     let config = NodejsConfig::try_load(module.config);
     let nodejs_version = utils::exec_cmd("node", &["--version"])?.stdout;
+    let _engines_version = get_engines_version(&context.current_dir);
+    let in_engines_range = check_engines_version(&nodejs_version, _engines_version);
```
As below
```suggestion
    let engines_version = get_engines_version(&context.current_dir);
    let in_engines_range = check_engines_version(&nodejs_version, engines_version);
```

andytom reviewed (commented):
Hi @t-mangoe thank you for your contribution. I have added a few suggestions hopefully nothing major.

Can you also update the Docs for the [NodeJS module](https://github.com/starship/starship/tree/master/docs/config#nodejs) to include this additional configuration (you only need to update the English docs all translations are done via [translate.starship.rs](https://translate.starship.rs/)).

t-mangoe commented:
@andytom Thank you for your review.
 Following your suggestions, I have fixed my codes.
 Also, I updated the Docs for the [NodeJS module](https://github.com/starship/starship/tree/master/docs/config#nodejs).

andytom commented:
> @andytom Thank you for your review.
> Following your suggestions, I have fixed my codes.
> Also, I updated the Docs for the [NodeJS module](https://github.com/starship/starship/tree/master/docs/config#nodejs).

Hi @t-mangoe, Sorry for the delay in replying. Thanks for fixing the issues unfortunately it looks like there are still some conflicts with the `Cargo.toml` if you can fix those conflicts I think we should be good to go here.

t-mangoe commented:
@andytom 
Oh, sorry. I overlooked the conflicts in Cargo.toml.
I resolved these conflicts.

t-mangoe commented:
I have resolved conflicts.

andytom commented:
Thank you for your contribution @t-mangoe
